### PR TITLE
feat(animations): allow animation integration support into host params

### LINKED ISSUE #9044: Animations cannot be set via `@host` or `Component.host`

### PR DESCRIPTION
**Pull request by matsko**
Closes #9044

### CODE REVIEW

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ export class AnimationCompiler {
         groupedErrors.push(errorMessage);
       }
 
-      var factoryName = `${component.type.name}_${entry.name}_${index}`;
+      var factoryName = `${component.type.name}_${entry.name}`;
```
Comment (matsko) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: Make sure this is computed in one place

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 2]
```diff
@@ -878,6 +881,19 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+         inject(
+             [TestComponentBuilder, AnimationDriver],
+             fakeAsync((tcb: TestComponentBuilder, driver: MockAnimationDriver) => {
+               makeAnimationCmp(
+                   tcb, '<div class="target" [@status]="exp"></div>', [], () => {}, (e: any) => {
+                     var message = e.message;
+                     expect(message).toMatch(
+                         /Animation parsing for DummyIfCmp has failed due to the following errors:/);
+                     expect(message).toMatch(/- couldn't find an animation entry for status/);
+                   });
```
Comment (JurajMlich) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: If the component compiles successfully, the error callback would not get called, am I right? And in that case the expect call wouldn't get invoked and the test would pass.

Comment (matsko) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: Good catch. The test should contain both callbacks with the first callback throwing an exception if it gets run.

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 3]
```diff
@@ -360,3 +375,49 @@ function _isEndStateAnimateStep(step: AnimationAst): boolean {
...
+        if (!isPresent(this._animationRegistry[animationName])) {
+          this.errors.push(
+              new AnimationParseError(`couldn't find an animation entry for ${animationName}`));
+        }
+      }
+    });
+    templateVisitAll(this, ast.children);
+  }
+
+  visitBoundText(ast: BoundTextAst, ctx: any): any { this._nodeIndex++; }
```
Comment (matsko) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: remove ++

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ import {BaseException} from '../facade/exceptions';
 import {isArray, isBlank, isPresent} from '../facade/lang';
 import {Identifiers} from '../identifiers';
 import * as o from '../output/output_ast';
+import {PropertyBindingType, TemplateAst, TemplateAstVisitor, NgContentAst, EmbeddedTemplateAst, ElementAst, ReferenceAst, VariableAst, BoundEventAst, BoundElementPropertyAst, AttrAst, BoundTextAst, TextAst, DirectiveAst, BoundDirectivePropertyAst, templateVisitAll,} from '../template_ast';
```
Comment (matsko) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: `* as t`

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 5]
```diff
@@ -44,12 +45,15 @@ function declareTests({useJit}: {useJit: boolean}) {
     var makeAnimationCmp =
         (tcb: TestComponentBuilder, tpl: string,
          animationEntry: AnimationEntryMetadata | AnimationEntryMetadata[],
-         callback: any /** TODO #9100 */ = null) => {
+         callback: Function = null, failure: Function = null) => {
```
Comment (matsko) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: Function param type `(fixture: any) => void`

File: modules/@angular/core/test/animation/animation_integration_spec.ts [hunk 6]
```diff
@@ -878,6 +882,24 @@ function declareTests({useJit}: {useJit: boolean}) {
...
+             [TestComponentBuilder, AnimationDriver],
+             fakeAsync((tcb: TestComponentBuilder, driver: MockAnimationDriver) => {
+               makeAnimationCmp(
+                   tcb, '<div class="target" [@status]="exp"></div>', [],
+                   () => {
+                     throw new BaseException(
+                         'Error: expected animations for DummyIfCmp to throw an error within this spec');
+                   },
+                   (e: any) => {
+                     var message = e.message;
```
Comment (matsko) [inline modules/@angular/core/test/animation/animation_integration_spec.ts]: const

File: modules/@angular/compiler/src/animation/animation_compiler.ts [hunk 7]
```diff
@@ -33,23 +33,33 @@ export class AnimationCompiler {
     var compiledAnimations: CompiledAnimation[] = [];
     var index = 0;
     var groupedErrors: string[] = [];
+    var triggerLookup: {[key: string]: CompiledAnimation} = {};
+    var componentName = component.type.name;
 
     component.template.animations.forEach(entry => {
       var result = parseAnimationEntry(entry);
+      var triggerName = entry.name;
```
Comment (matsko) [inline modules/@angular/compiler/src/animation/animation_compiler.ts]: const

File: modules/@angular/compiler/test/animation/animation_compiler_spec.ts [hunk 8]
```diff
@@ -57,5 +64,23 @@ export function main() {
...
+
+    it('should throw an error when two or more animation triggers contain the same name', () => {
+      var doCompile = () => {
+        var t1Data: any[] = [];
+        var t2Data: any[] = [];
+        compileTriggers([['myTrigger', t1Data], ['myTrigger', t2Data]]);
+      };
+
+      var capturedErrorMessage: string;
+      try {
```
Comment (matsko) [inline modules/@angular/compiler/test/animation/animation_compiler_spec.ts]: expect toThrow

**Comment by matsko** [comment]
LGTM from @vicb

**Pull request closed without merging**